### PR TITLE
 feat: call owncloud api to create share link

### PR DESCRIPTION
**Pull request by Iloniavo**

### CODE REVIEW

File: doc/api.yml [hunk 1]
```diff
@@ -3825,6 +3859,19 @@ components:
...
+        expiration:
+          type: string
+        name:
+          type: string
+        path:
+          type: string
+        url:
+          type: string
+        permissions:
+          type: string
```
Comment (Mahefaa) [inline doc/api.yml]: permissions ? means it's plural ? 
if so, it has to be an array of permissions.

If permissions are known already known, you should use an ENUM

File: src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package school.hei.haapi.endpoint;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java]: I don't think this is a good idea to put in endpoint package, it is more of a service.owncloud file than an endpoint one

File: src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  public OwnCloudConf(
+      @Value("${OWNCLOUD_BASE_URL}") String baseUrl,
+      @Value("${OWNCLOUD_USERNAME}") String username,
+      @Value("${OWNCLOUD_PASSWORD}") String password) {
+    this.baseUrl =
+        UriComponentsBuilder.fromHttpUrl(baseUrl)
+            .path("/ocs/v1.php/apps/files_sharing/api/v1/shares")
+            .queryParam("path={path}")
+            .queryParam("name={name}")
+            .queryParam("shareType=3")
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java]: 3 is a magic number

File: src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Map;
+import lombok.Getter;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.util.UriComponents;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Configuration
+public class OwnCloudConf {
+  private final UriComponents baseUrl;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java]: rename baseUrl to fileTemplateUrl

File: src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            .queryParam("expireDate={expireDate}")
+            .queryParam("format=json")
+            .queryParam("publicUpload=false")
+            .build();
+    this.username = username;
+    this.password = password;
+  }
+
+  public URI getURI(String path, String name, Integer permissions) {
+    Map<String, Object> uriVariables =
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/OwnCloudConf.java]: use Map<String, String> and format permissions and Instant in the Map

File: src/main/java/school/hei/haapi/endpoint/RestTemplateConf.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.boot.web.client.RestTemplateBuilder;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.web.client.ResponseErrorHandler;
+import org.springframework.web.client.RestTemplate;
+import school.hei.haapi.model.exception.ApiException;
+
+@Configuration
+@Slf4j
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/RestTemplateConf.java]: rm slf4j

File: doc/api.yml [hunk 7]
```diff
@@ -2451,6 +2451,40 @@ paths:
           $ref: '#/components/responses/429'
         '500':
           $ref: '#/components/responses/500'
+  '/school/files/share_link':
+    post:
```
Comment (Mahefaa) [inline doc/api.yml]: i assume that "POST" method was chosen because every call to this api will create a new link ?

File: doc/api.yml [hunk 8]
```diff
@@ -2451,6 +2451,40 @@ paths:
...
+        - name: name
+          in: query
+          required: true
+          schema:
+            type: string
+        - name: path
+          in: query
+          required: true
+          schema:
+            type: string
```
Comment (Mahefaa) [inline doc/api.yml]: add a description for both.
also add a pattern ( should path start with '/' ? should path contain '/', what about the name ?, can it contain spaces ?)

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FileInfoMapper.java [hunk 9]
```diff
@@ -24,4 +28,20 @@ public school.hei.haapi.endpoint.rest.model.FileInfo toRest(FileInfo fileInfo) {
         .fileUrl(presignedUrl)
         .creationDatetime(fileInfo.getCreationDatetime());
   }
+
+  public ShareInfo toShareInfo(OcsData ocsData) {
+    log.info("sssssss");
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FileInfoMapper.java]: rm

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FileInfoMapper.java [hunk 10]
```diff
@@ -24,4 +28,20 @@ public school.hei.haapi.endpoint.rest.model.FileInfo toRest(FileInfo fileInfo) {
...
+    return new ShareInfo()
+        .name(ocsData.getName())
+        .path(ocsData.getPath())
+        .url(ocsData.getUrl())
+        .permissions(
+            switch (ocsData.getPermissions()) {
+              case 1 -> "READ";
+              case 2 -> "READ/WRITE";
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FileInfoMapper.java]: hah finally some use of my Effective Java Knowledge.

create an enum named OwnCloudFilePermissions, with the needed values, then return an EnumSet in the permissions attribute

File: src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java [hunk 11]
```diff
@@ -143,6 +143,8 @@ req, res, null, forbiddenWithRemoteInfo(req))))
                     .hasAnyRole(MANAGER.getRole(), TEACHER.getRole(), STUDENT.getRole())
                     .requestMatchers(GET, "/school/files/*")
                     .hasAnyRole(MANAGER.getRole(), STUDENT.getRole(), TEACHER.getRole())
+                    .requestMatchers(POST, "/school/files/share_link")
+                    .hasAnyRole(MANAGER.getRole(), STUDENT.getRole(), TEACHER.getRole())
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/security/SecurityConf.java:147]: info: what are we supposed to do with "non school-files" ? 
( this is a non-blocking comment, I would simply like to get more informations)

File: src/main/java/school/hei/haapi/service/OwnCloudService.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        switch (currentUser.getRole()) {
+          case STUDENT, TEACHER -> 1;
+          case MANAGER -> 31;
+        };
+
+    log.info(conf.getUsername(), conf.getPassword());
+    HttpHeaders headers = new HttpHeaders();
+    String auth = conf.getUsername() + ":" + conf.getPassword();
+    String authHeader =
+        "Basic " + new String(Base64.getEncoder().encode(auth.getBytes(StandardCharsets.US_ASCII)));
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/OwnCloudService.java]: static import US_ASCII

Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/OwnCloudService.java]: create a method computing the authorization header from username and password

File: src/main/java/school/hei/haapi/service/OwnCloudService.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    headers.set("Authorization", authHeader);
+    HttpEntity<Void> entity = new HttpEntity<>(null, headers);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/OwnCloudService.java]: create a method computing Http Headers from needed values ( username, password)

File: src/main/java/school/hei/haapi/service/OwnCloudService.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+    try {
+      ResponseEntity<String> response =
+          restTemplate.exchange(conf.getURI(path, name, permission), POST, entity, String.class);
+      JsonNode node = objectMapper.readTree(response.getBody()).get("ocs").get("data");
+
+      return objectMapper.convertValue(node, OcsData.class);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/OwnCloudService.java]: I strongly believe that you can put the response extraction from some sort of RestTemplate Response Extractor or Response Converter

File: src/test/java/school/hei/haapi/integration/SchoolFileIT.java [hunk 15]
```diff
@@ -37,15 +38,75 @@
...
+          }""";
+
+  @Test
+  void manager_get_share_link() throws ApiException {
+    ApiClient apiClient = anApiClient(MANAGER1_TOKEN);
+    FilesApi filesApi = new FilesApi(apiClient);
+
+    ShareInfo actual = filesApi.createShareLink("name", "/Test-api");
+    log.info(actual.toString());
+    //    assertEquals(actual.getUrl(), "https://owncloud.server.mock");
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/SchoolFileIT.java]: incomplete test

File: src/test/java/school/hei/haapi/integration/conf/AbstractContextInitializer.java [hunk 16]
```diff
@@ -9,6 +9,8 @@ public abstract class AbstractContextInitializer
     implements ApplicationContextInitializer<ConfigurableApplicationContext> {
   private static final String SENTRY_MOCK_DSN = "https://examplePublicKey@example.sentry.io/12345";
 
+  public AbstractContextInitializer() {}
+
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/conf/AbstractContextInitializer.java]: rm this NoArgsConstructor

File: src/main/java/school/hei/haapi/service/OwnCloudService.java (hunk 14 above)
Comment (Mahefaa) [reply src/main/java/school/hei/haapi/service/OwnCloudService.java]: it is done whether by correctly designing the OCSResponse class or by using 
RestTemplate::execute method

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: doc/api.yml (hunk 7 above)
Comment (Iloniavo) [reply doc/api.yml]: yes

File: src/main/java/school/hei/haapi/service/ownCloud/OwnCloudService.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+  public OcsData createShareLink(String path) {
+    User currentUser = getPrincipal().getUser();
+    Integer permission =
+        switch (currentUser.getRole()) {
+          case STUDENT, TEACHER -> 1;
+          case MANAGER -> 15;
+        };
+
+    log.info("User");
+    log.info(currentUser.toString());
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/ownCloud/OwnCloudService.java]: do not forget to rm these logs

File: src/main/java/school/hei/haapi/service/utils/OwnCloudUtils.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import lombok.Getter;
+import school.hei.haapi.endpoint.rest.model.OwnCloudPermission;
+
+@Getter
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/utils/OwnCloudUtils.java]: useless Getter

**Review by Mahefaa — CHANGES_REQUESTED** [review]

**Pull request merged**